pkg/tsdb/cloudwatch/response_parser.go

```diff
@@ -7,11 +7,85 @@ import (
 	"strings"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/gigawattio/awsarn"
 	"github.com/grafana/grafana/pkg/components/null"
 	"github.com/grafana/grafana/pkg/tsdb"
 )
 
+/* See https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_actions-resources-contextkeys.html for service resource type tables.
+ * See https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/aws-services-cloudwatch-metrics.html for CloudWatch namespaces.
+ */
+var resourceTypesMap = map[string][]string{
+	"AWS/EC2":            {"ec2:instance"},
+	"AWS/CloudFront":     {"cloudfront:distribution"},
+	"AWS/ApplicationELB": {"elasticloadbalancing:loadbalancer/app", "elasticloadbalancing:targetgroup"},
+	"AWS/ECS":            {"ecs:cluster", "ecs:service"},
+	"AWS/RDS":            {"rds:db"},
+	"AWS/S3":             {"s3"},
+}
+
+var resourcePrefixesMap = map[string]string{
+	// ELBv2 target groups have 'targetgroup/' prepended in the label but not the resource.
+	"elasticloadbalancing:targetgroup": "targetgroup/",
+}
+
+var resourceComponentMap = map[string]int{
+	// ECS services have the cluster name prepended in the resource but not the label.
+	"ecs:service": 1,
+}
+
+func (e *cloudWatchExecutor) getMetricResourceTags(queries map[string]*cloudWatchQuery) (map[string]map[string]string, error) {
+	/* We can query multiple resource types simultaneously, but only one region at a time, so we collect all query resource types into
+	 * regional buckets.
+	 */
+	resourceTypesByRegion := map[string][]string{}
+	for _, query := range queries {
+		if resources, ok := resourceTypesMap[query.Namespace]; ok {
+			resourceTypesByRegion[query.Region] = append(resourceTypesByRegion[query.Region], resources...)
+		}
+	}
+
+	tags := map[string]map[string]string{}
+	for region, resourceTypes := range resourceTypesByRegion {
+		typeFilters := []*string{}
+		typeFiltersAdded := map[string]bool{}
+		for _, tp := range resourceTypes {
+			if _, ok := typeFiltersAdded[tp]; !ok {
+				typeFilters = append(typeFilters, aws.String(tp))
+				typeFiltersAdded[tp] = true
+			}
+		}
+
+		resources, err := e.resourceGroupsGetResources(region, nil, typeFilters)
```

Comment:
What's the performance/cost of doing this call per each query/region? At a first glance seems like it could be problematic?

Reply:
Queries in the same region are consolidated into the same RGTA query, so for most panels (which are single region) this will just be a single RGTA call.

RGTA is a very lightweight API (basically just key/value fetching), so the performance impact is negligible.

Additionally, the existing "solution" for this sort of lookup (at least for EC2 instances) is via ec2_instance_attribute at the dashboard level with query variables, which just boils down to a bunch of RGTA queries in the background anyway (one per instance, rather than one per panel).